### PR TITLE
Don't consider receiving non-OK status as an error for HTTP2

### PR DESCRIPTION
**Pull request by yashykt**
TL;DR - Clients that depend on the behavior of the transports dropping messages for non-OK statuses would get broken by this change.

Currently, gRPC Core transports drop unread messages at the transport layer if it receives a non-OK status while it still has pending messages to report to the application. More formally, all the following conditions need to be satisfied for the transports to drop messages -
1) The transport has read one or more messages (data frames for HTTP2) and has not yet forwarded the messages to the application (possibly because the application has not yet read them).
2) The server has ended the RPC with a non-OK status.
3) The op GRPC_OP_RECV_STATUS_ON_CLIENT has been sent to gRPC Core before all the messages have been read by the application.

This behavior essentially treats the non-OK status as an RPC error, thereby discarding pending messages. This is unwanted behavior and the preferred behavior would be to just treat the non-OK status as an OK status at the transport layer so that the application can see all messages before the RPC ends.

Even though this is a fix, there would still be a behavior change for clients. Hence, clients that depend on the behavior of the transports dropping messages for non-OK statuses would get broken by this change. They additionally run the risk of causing a deadlock due to flow control. Applications that properly make sure that there are no more messages to be read either explicitly through a failing GRPC_OP_RECV_MESSAGE or implicitly through application guarantees will be unaffected.

Note that this PR makes this change only for the HTTP2 transport. A similar fix would need to be made for inproc and cronet transports.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/19545)

### CODE REVIEW

**Comment by markdroth** [comment]
Is this actually safe?  I thought we decided that it wasn't safe to do this yet, because if the client never reads but is expecting status, it could lead to a deadlock in the case where there is not enough flow control window to receive the trailing metadata.

**Comment by yashykt** [comment]
@markdroth I agree that this does not prevent the possible deadlock, but it solves the immediate issue on losing streaming messages for the callback api on non-ok status. There are test failures, so i need to check what's going wrong though

**Comment by yashykt** [comment]
also, @vjpai mentioned that we don't want to add a C++ async api test for this since our documentation prohibits this. I will either add a C core test or a callback api test

**Comment by yashykt** [comment]
@markdroth @vjpai ping?

**Comment by markdroth** [comment]
The updated PR description for the release notes looks pretty good.  Can you also explicitly point out that applications that do not read until they get a not-ok result risk causing a deadlock due to flow control?  Thanks!

**Comment by yashykt** [comment]
> The updated PR description for the release notes looks pretty good. Can you also explicitly point out that applications that do not read until they get a not-ok result risk causing a deadlock due to flow control? Thanks!

If the applications do not read until they get a result, they risk a deadlock irrespective of whether the status was ok or not, since the server could be potentially unable to send further messages waiting on flow control, blocking the status too.

**Comment by yashykt** [comment]
___
*[test/core/end2end/tests/streaming_error_response.cc, line 225 at r3](https://reviewable.io/reviews/grpc/grpc/19545#-MDq0AyOB9s9HO2jPb3G:-MDvDkvl7tJpwXNKDz_Z:b-sqcq2t) ([raw file](https://github.com/grpc/grpc/blob/6f55e8e6fe903a0bde1e80206fac89a4dbdcb097/test/core/end2end/tests/streaming_error_response.cc#L225)):*
> How did you trigger the server-side cancellation?  If you do it with grpc_call_cancel()
Yes

**Comment by yashykt** [comment]
Thanks for reviewing!
